### PR TITLE
Core/Players: Update the player's zone and area only after update the player's position server-side

### PR DESCRIPTION
**Pull request by schell244**
## Changes Proposed:
Port player zone and area update logic from TC.
I was looking into player map explore code and found this commit while comparing with TC. 

Sadly this does not fix the map explore issue after login mentioned in the comment of https://github.com/azerothcore/azerothcore-wotlk/issues/7737

## Issues Addressed:
- 

## SOURCE:
https://github.com/TrinityCore/TrinityCore/pull/12171

## Tests Performed:
builds, checked ingame

## How to Test the Changes:
1. create new character
2. enter new areas and zone
3. check map explore, channel messages, unmount in buildings, ...

## Known Issues and TODO List:
- 

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by Nyeriah — APPROVED** [review]

**Comment by claudiodfc** [comment]
## Test result 🦗

No changes after the PR:
image

**Comment by claudiodfc** [comment]
The relog issue still remains, however the chat messages and discovery states are still ok.

File: src/server/game/Entities/Player/Player.h [hunk 1]
```diff
@@ -1805,6 +1805,7 @@ class Player : public Unit, public GridObject<Player>
     void UpdatePvP(bool state, bool _override = false);
     void UpdateZone(uint32 newZone, uint32 newArea);
     void UpdateArea(uint32 newArea);
+    void SetNeedsZoneUpdate(bool needsUpdate) {m_needsZoneUpdate = needsUpdate;}
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/Player.h:1808]:
```suggestion
    void SetNeedZoneUpdate(bool needUpdate) { m_needsZoneUpdate = needsUpdate; }
```

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -194,6 +194,8 @@ Player::Player(WorldSession* session): Unit(true), m_mover(this)
     m_areaUpdateId = 0;
     m_team = TEAM_NEUTRAL;
 
+    m_needsZoneUpdate = false;
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/Player.cpp:197]:
```suggestion
    m_needZoneUpdate = false;
```

File: src/server/game/Entities/Player/Player.h [hunk 3]
```diff
@@ -2868,6 +2869,8 @@ class Player : public Unit, public GridObject<Player>
 
     uint8 m_grantableLevels;
 
+    bool m_needsZoneUpdate;
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/Player.h:2872]:
```suggestion
    bool m_needZoneUpdate;
```

File: src/server/game/Entities/Player/PlayerUpdates.cpp [hunk 4]
```diff
@@ -1115,6 +1115,15 @@ bool Player::UpdatePosition(float x, float y, float z, float orientation,
     if (!Unit::UpdatePosition(x, y, z, orientation, teleport))
         return false;
 
+    // Update player zone if needed
+    if (m_needsZoneUpdate)
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/PlayerUpdates.cpp:1119]:
```suggestion
    if (m_needZoneUpdate)
```

File: src/server/game/Entities/Player/PlayerUpdates.cpp [hunk 5]
```diff
@@ -1115,6 +1115,15 @@ bool Player::UpdatePosition(float x, float y, float z, float orientation,
...
+    {
+        uint32 newZone, newArea;
+        GetZoneAndAreaId(newZone, newArea);
+        UpdateZone(newZone, newArea);
+        m_needsZoneUpdate = false;
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/PlayerUpdates.cpp:1124]:
```suggestion
        m_needZoneUpdate = false;
```

File: src/server/game/Handlers/MiscHandler.cpp [hunk 6]
```diff
@@ -530,10 +530,9 @@ void WorldSession::HandleZoneUpdateOpcode(WorldPacket& recv_data)
 
     LOG_DEBUG("network", "WORLD: Recvd ZONE_UPDATE: %u", newZone);
 
-    // use server size data
-    uint32 newzone, newarea;
-    GetPlayer()->GetZoneAndAreaId(newzone, newarea);
-    GetPlayer()->UpdateZone(newzone, newarea);
+    // use server side data, but only after update the player position. See Player::UpdatePosition().
+    GetPlayer()->SetNeedsZoneUpdate(true);
```
Comment (Kitzunu) [inline src/server/game/Handlers/MiscHandler.cpp:534]:
```suggestion
    GetPlayer()->SetNeedZoneUpdate(true);
```

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Player/Player.h [hunk 7]
```diff
@@ -1805,6 +1805,7 @@ class Player : public Unit, public GridObject<Player>
     void UpdatePvP(bool state, bool _override = false);
     void UpdateZone(uint32 newZone, uint32 newArea);
     void UpdateArea(uint32 newArea);
+    void SetNeedZoneUpdate(bool needUpdate) { m_needsZoneUpdate = needsUpdate; }
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/Player.h:1808]:
```suggestion
    void SetNeedZoneUpdate(bool needUpdate) { m_needZoneUpdate = needUpdate; }
```

**Review by Kitzunu — CHANGES_REQUESTED** [review]
forgot about these lol

**Comment by schell244** [comment]
Thanks for checking :)

**Review by Kitzunu — APPROVED** [review]

**Pull request merged**